drivers: clock: rcar: Rework structure & filetree (Using Soc directory)

aaillet opened the pull request:
This PR is reworking file tree for clocks definitions for renesas rcar SoCs.

Goals of these commits are:
  - to deploy a clean clocks definitions file tree as close as linux one as possible.
  - to use same clock names as linux ones for unification purpose.
  - to deploy a file tree and rework the drive so that it will be easier to add new SoCs later.

Summary of changes:
- dt-bindings: rework renesas clocks dt-bindings
  - regroup renesas related dt-bindings in a folder
  - rename renesas rcar common dt-binding to match linux name
  - add soc specific dt-binding matching linux names
  - soc dt-bindings are defining clocks matching linux names

- dts: use new clocks names
  - move clocks definitions in SoC layer for each core clock entry

- soc: move some part of clock driver into SoC specific files
  - clock_rcar header defines prototypes for all vendor boards (renesas)
  - clock soc header defines common values for a whole family (gen3)
  - clock_r8a.. source is implementing soc specific methods and registers

- driver: use new clocks names
  - call for SoC defined methods
  - driver will not grow exponentially with number of SoCs to support
  - most of the new changes should now be in SoC files

These changes have been tested on r8a77951 SoC's powered H3ULCB board.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
dts: rcar: Rework clock definition

This commit is reworking clock definition in
order to match linux filetree and definition
as much as possible.

- dt-bindings: rework renesas clocks dt-bindings
  - regroup renesas related dt-bindings in a folder
  - rename renesas rcar common dt-binding to match linux name
  - add soc specific dt-binding matching linux names
  - soc dt-bindings are defining clocks matching linux names

- dts: use new clocks names
  - move clocks definitions in SoC layer for each core clock entry

- driver: use new clocks names

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
drivers: clock: rcar: Deploy soc defined methods

This commit rewrite renesas R-Car clock driver in order
to be able to support any new SoC later without
letting it grow exponentially.

This work is so moving some logic in soc files rather than
letting them in the driver itself.

- soc: move some part of clock driver into SoC specific files
  - clock_rcar header defines prototypes for all vendor boards (renesas)
  - clock soc header defines common values for a whole family (gen3)
  - clock_r8a.. source is implementing soc specific methods and registers

- drivers: call for SoC defined methods
  - driver will not grow exponentially with number of SoCs to support
  - every new edit should now be in SoC files

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

julien-massot commented on include/dt-bindings/clock/renesas/r8a77951-cpg-mssr.h:
```diff
@@ -0,0 +1,68 @@
+#define R8A7795_CLK_S3D4		18      /* SCIF clock */
+#define R8A7795_CLK_LB			19
+#define R8A7795_CLK_CL			20
+#define R8A7795_CLK_ZB3		21
```
please fix indentation here

aaillet replied:
fixed it, thanks

julien-massot commented on soc/arm/renesas_rcar/gen3/CMakeLists.txt:
```diff
@@ -4,4 +4,5 @@
 )
-zephyr_library_sources_ifdef(CONFIG_SOC_R8A77951 pfc_r8a77951.c)
+
+zephyr_library_sources_ifdef(CONFIG_SOC_R8A77951 pfc_r8a77951.c clock_r8a77951.c)
```
cpg_r8a77951.c ?

aaillet replied:
renamed the file, thanks for the hint

julien-massot commented on soc/arm/renesas_rcar/common/clock_rcar.h:
```diff
@@ -0,0 +1,19 @@
+
+#include <dt-bindings/clock/renesas/renesas-cpg-mssr.h>
+
+int soc_cpg_core_clock_endisable(uint32_t base_address, uint32_t module,
```
Please add documentation about these functions which include the possible error code.

aaillet replied:
just documented these methods !

julien-massot commented on soc/arm/renesas_rcar/gen3/clock_r8a77951.c:
```diff
@@ -0,0 +1,126 @@
+	uint32_t bit = module % 100;
+
+	if (bit > 32 || reg > ARRAY_SIZE(rmstpcr)) {
+		return 1;
```
-EINVAL ?

julien-massot commented:
Thanks Aymeric, this is already in a good shape and will make easier to introduce future SoC series.

julien-massot commented on soc/arm/renesas_rcar/gen3/cpg_r8a77951.c:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	return srcr[reg];
+}
```
So in one case soc_cpg_get_mstpcr_offset we take a module (reg_idx + bit) and in another casesoc_cpg_get_srcr_offset we take a reg_idx directly, my feeling is that we should go for reg_idx in both cases.

aaillet replied:
the two functions have now much closer prototypes, their functioning and returns are similar

julien-massot commented on soc/arm/renesas_rcar/gen3/cpg_r8a77951.c:
```diff
@@ -0,0 +1,130 @@
+uint16_t soc_cpg_get_srcr_offset(uint32_t reg)
+{
+	if (reg > ARRAY_SIZE(srcr)) {
+		return -EINVAL;
```
the return value is -EINVAL but the functions returns an uint16_t

aaillet replied:
new prototype changes return value to int

julien-massot commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -59,71 +29,37 @@ static void cpg_write(const struct rcar_mssr_config *config,
+		irq_unlock(key);
+		return 1;
+	}
+
```
you can lock the irq after getting the register offset. please also return the value of soc_cpg_get_mstpcr_offset in case of error.

int ret;
unsigned int key;

ret = soc_cpg_get_mstpcr_offset(module, &offset);
if (ret < 0) {
return ret;
}
key = irq_lock();

julien-massot commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -59,71 +29,37 @@ static void cpg_write(const struct rcar_mssr_config *config,
-	sys_write32(reg_val, config->base_address + RMSTPSR(reg));
+	sys_write32(reg_val, config->base_address + offset);
 	if (!enable) {
 		cpg_reset(config, reg, bit);
```
we may simply drop the cpg_reset call and remove cpg_reset implementation until we have a reset framework in zephyr.

julien-massot commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -56,76 +26,60 @@ static void cpg_write(const struct rcar_mssr_config *config,
 	if (!enable) {
-		cpg_reset(config, reg, bit);
+		ret = cpg_reset(config, reg, bit);
+		if (ret < 0) {
```
you can initialise ret to 0
and just return ret at the end of the function, so that you don't have to call irq_unlock

aaillet replied:
As we are returning from this method if `soc_cpg_get_mstpcr_offset` is returning an error, ret should still be at 0 if we reach `cpg_reset call`, so it's a good way to clean the code a bit. thanks

julien-massot commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -170,30 +123,13 @@ static int cpg_get_rate(const struct device *dev,
-	}
+	ret = soc_cpg_get_rate(config->base_address, clk->module, rate);
 
 	return ret;
```
return soc_cpg_get_rate(config->base_address, clk->module, rate); ?

aaillet replied:
fixed it :+1:

aaillet commented:
Just resolved a conflict after merge of #43056.

gmarull commented on include/dt-bindings/clock/renesas/renesas-cpg-mssr.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021 IoT.bzh
+ * Copyright (c) 2022 IoT.bzh
```
nit: `2021-2022`

aaillet replied:
fixed copyrights, thanks

gmarull commented on soc/arm/renesas_rcar/common/clock_rcar.h:
```diff
@@ -0,0 +1,63 @@
+ * @retval 0 on success
+ * @retval -EINVAL if clock or rate not supported
+ */
+int soc_cpg_core_clock_endisable(uint32_t base_address, uint32_t module,
```
why is this code moved under `soc/`? why can't it be under `drivers/clock_control` (I guess driver can have its own folder with multiple sources if it helps).

aaillet replied:
Hi @gmarull, thanks for your review ! 
I moved these part of code into `/soc` for multiples reasons:

- By creating a driver file for each SoC, this `/driver/clock_control/renesas` folder (and so Zephyr sources) could drastically grow in a way that could be hard to maintain.
Each new SoC added (3 SoCs are already on the way) would add their own clock driver that would duplicate a lot of common code with other SoC ones.

- Indeed, especially for renesas SoCs, clocks trees are SoC specific, so, when we will add new SoC in the future, it will be easier to add the associated implementation in the newly created `/soc` folder instead of doing changes to drivers. Also, the git history would be much cleaner this way (by not editing both `/soc` and `/driver` domain).

- This rework and the choice of the `/soc` directory also follows the work concerning the pinctrl driver for Renesas rcar SoCs (#40550), I copied the way that this implementation is creating a source code per SoC in the `/soc` directory to implement SoC specific part of the code for the sake of homogeneity. 

- The hierarchy created with clock_rcar.h (renesas) -> clock_soc.h (gen3) -> cpg_r8a..c (soc) files is also allowing us to duplicate the least possible code in the Zephyr code base. This scheme would not be as clean as it is if moved under the `/driver` folder.

julien-massot replied:
We (Aymeric :) ) will give a try with a reversed logic.
One driver per SoC under a drivers/clock_control/renesas directory, which will be enabled based on SoC config value.
that way the current clock_control_rcar_cpg_mssr.c will export some common function that will be used by the different driver.

carlescufi commented:
@julien-massot @nordic-krch  could you please take a look?

julien-massot commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -1,52 +1,22 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
 #define DT_DRV_COMPAT renesas_rcar_cpg_mssr
```
please change this file to export common function that will be used by each renesas soc specific driver.

aaillet commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -56,109 +26,92 @@ static void cpg_write(const struct rcar_mssr_config *config,
 	} else if (clk->domain == CPG_CORE) {
-		ret = cpg_core_clock_endisable(dev, clk->module, 0, false);
+		ret = soc_cpg_core_clock_endisable(config->base_address, clk->module,
+						   clk->rate, true);
```
```suggestion
						   clk->rate, false);
```

aaillet commented on drivers/clock_control/clock_control_rcar_cpg_mssr.c:
```diff
@@ -56,109 +26,92 @@ static void cpg_write(const struct rcar_mssr_config *config,
 
-unlock:
-	irq_unlock(key);
 	return 0;
```
```suggestion
	return ret;
```

aaillet committed:
drivers: clock: rcar: Rename global includes file

Rename r-car clock driver global include file
in order to match other files names.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet commented:
New version deploying a driver per soc in /driver folder is available in #44418.
The version from this PR is therefore obsolete.
